# Загрузка Pandas и очистка данных

In [205]:
import pandas as pd

In [206]:
df = pd.read_csv('main_task.csv')

In [207]:
df.head()

Restaurant_id       City                                      Cuisine Style  \
0       id_5569      Paris            ['European', 'French', 'International']   
1       id_1535  Stockholm                                                NaN   
2        id_352     London  ['Japanese', 'Sushi', 'Asian', 'Grill', 'Veget...   
3       id_3456     Berlin                                                NaN   
4        id_615     Munich  ['German', 'Central European', 'Vegetarian Fri...   

   Ranking  Rating Price Range  Number of Reviews  \
0   5570.0     3.5    $$ - $$$              194.0   
1   1537.0     4.0         NaN               10.0   
2    353.0     4.5        $$$$              688.0   
3   3458.0     5.0         NaN                3.0   
4    621.0     4.0    $$ - $$$               84.0   

                                             Reviews  \
0  [['Good food at your doorstep', 'A good hotel ...   
1  [['Unique cuisine', 'Delicious Nepalese food']...   
2  [['Catch up with friends', 'Not exceptional'],...   
3                                           [[], []]   
4  [['Best place to try a Bavarian food', 'Nice b...   

                                              URL_TA     ID_TA  
0  /Restaurant_Review-g187147-d1912643-Reviews-R_...  d1912643  
1  /Restaurant_Review-g189852-d7992032-Reviews-Bu...  d7992032  
2  /Restaurant_Review-g186338-d8632781-Reviews-RO...  d8632781  
3  /Restaurant_Review-g187323-d1358776-Reviews-Es...  d1358776  
4  /Restaurant_Review-g187309-d6864963-Reviews-Au...  d6864963

In [208]:
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 40000 entries, 0 to 39999
Data columns (total 10 columns):
 #   Column             Non-Null Count  Dtype  
---  ------             --------------  -----  
 0   Restaurant_id      40000 non-null  object 
 1   City               40000 non-null  object 
 2   Cuisine Style      30717 non-null  object 
 3   Ranking            40000 non-null  float64
 4   Rating             40000 non-null  float64
 5   Price Range        26114 non-null  object 
 6   Number of Reviews  37457 non-null  float64
 7   Reviews            40000 non-null  object 
 8   URL_TA             40000 non-null  object 
 9   ID_TA              40000 non-null  object 
dtypes: float64(3), object(7)
memory usage: 3.1+ MB


Приведем имена колонок к нормальному виду

In [209]:
df.columns=['restaurant_id', 'city', 'cuisine_style', 'ranking', 'rating',
       'price_range', 'reviews_cnt', 'reviews', 'ta_url', 'ta_id']


### Общие наблюдения и очистки

In [210]:
df.describe(include='all')

restaurant_id    city cuisine_style       ranking        rating  \
count          40000   40000         30717  40000.000000  40000.000000   
unique         11909      31          9007           NaN           NaN   
top           id_436  London   ['Italian']           NaN           NaN   
freq              18    5757          1032           NaN           NaN   
mean             NaN     NaN           NaN   3676.028525      3.993037   
std              NaN     NaN           NaN   3708.749567      0.668417   
min              NaN     NaN           NaN      1.000000      1.000000   
25%              NaN     NaN           NaN    973.000000      3.500000   
50%              NaN     NaN           NaN   2285.000000      4.000000   
75%              NaN     NaN           NaN   5260.000000      4.500000   
max              NaN     NaN           NaN  16444.000000      5.000000   

       price_range   reviews_cnt   reviews  \
count        26114  37457.000000     40000   
unique           3           NaN     33516   
top       $$ - $$$           NaN  [[], []]   
freq         18412           NaN      6471   
mean           NaN    124.825480       NaN   
std            NaN    295.666352       NaN   
min            NaN      2.000000       NaN   
25%            NaN      9.000000       NaN   
50%            NaN     33.000000       NaN   
75%            NaN    115.000000       NaN   
max            NaN   9660.000000       NaN   

                                                   ta_url     ta_id  
count                                               40000     40000  
unique                                              39980     39980  
top     /Restaurant_Review-g187514-d987900-Reviews-Com...  d1315077  
freq                                                    2         2  
mean                                                  NaN       NaN  
std                                                   NaN       NaN  
min                                                   NaN       NaN  
25%                                                   NaN       NaN  
50%                                                   NaN       NaN  
75%                                                   NaN       NaN  
max                                                   NaN       NaN

Есть неуникальные _restaurant_id_ и _ta_id_

In [217]:
df[df.restaurant_id==df.restaurant_id.value_counts().index[0]].head(3)



restaurant_id        city  \
2955        id_436   Barcelona   
4674        id_436  Copenhagen   
6118        id_436      Oporto   

                                          cuisine_style  ranking  rating  \
2955  ['Seafood', 'Mediterranean', 'Spanish', 'Glute...    437.0     4.5   
4674  ['European', 'Spanish', 'Mediterranean', 'Vege...    437.0     4.0   
6118  ['European', 'Portuguese', 'Soups', 'Cafe', 'V...    437.0     4.5   

     price_range  reviews_cnt  \
2955    $$ - $$$        546.0   
4674    $$ - $$$         89.0   
6118           $         71.0   

                                                reviews  \
2955  [['Fresh Fish', 'Perfect concept'], ['01/03/20...   
4674  [['Lovely restaurant', 'Great atmosphere and g...   
6118  [['Super nice healthy and good', 'Great lunch ...   

                                                 ta_url     ta_id  
2955  /Restaurant_Review-g187497-d1592613-Reviews-La...  d1592613  
4674  /Restaurant_Review-g189541-d8738362-Reviews-To...  d8738362  
6118  /Restaurant_Review-g189180-d9463122-Reviews-Ca...  d9463122

Одинаковый _restaurant_id_ встречается в разных городах и имеет разный _ta_id_. Удалим эту колонку.

In [221]:
df.drop(["restaurant_id"], axis=1, inplace=True)

In [231]:
with pd.option_context('display.max_colwidth', -1):
    print(df[df.ta_id==df.ta_id.value_counts().index[0]].head(3))

city  \
10439  Madrid   
25785  Madrid   

                                                                    cuisine_style  \
10439  ['Mediterranean', 'Spanish', 'Vegetarian Friendly', 'Gluten Free Options']   
25785  ['Mediterranean', 'Spanish', 'Vegetarian Friendly', 'Gluten Free Options']   

       ranking  rating price_range  reviews_cnt  \
10439  469.0    4.0     $$ - $$$    355.0         
25785  512.0    4.0     $$ - $$$    355.0         

                                                                                       reviews  \
10439  [['Very good all around', 'Our first Madrid restaurant'], ['11/20/2017', '10/11/2017']]   
25785  [['Very good all around', 'Our first Madrid restaurant'], ['11/20/2017', '10/11/2017']]   

                                                                          ta_url  \
10439  /Restaurant_Review-g187514-d1315077-Reviews-El_Pimiento_Verde-Madrid.html   
25785  /Restaurant_Review-g187514-d1315077-Reviews-El_Pimiento_Verde-Madrid.html   


Судя по-всему дублирующиеся _ta_id_ отличаются только по _ranking_. Попробуем сгрупировать, при этом ранк возьмется случайно, что не страшно, т.к. дублей всего 20

In [237]:
df.drop_duplicates(subset=['city', 'cuisine_style', 'rating','price_range', 'reviews_cnt', 'reviews', 'ta_url', 'ta_id'], inplace=True, ignore_index=True)
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 39980 entries, 0 to 39979
Data columns (total 9 columns):
 #   Column         Non-Null Count  Dtype  
---  ------         --------------  -----  
 0   city           39980 non-null  object 
 1   cuisine_style  30701 non-null  object 
 2   ranking        39980 non-null  float64
 3   rating         39980 non-null  float64
 4   price_range    26101 non-null  object 
 5   reviews_cnt    37437 non-null  float64
 6   reviews        39980 non-null  object 
 7   ta_url         39980 non-null  object 
 8   ta_id          39980 non-null  object 
dtypes: float64(3), object(6)
memory usage: 2.7+ MB


Теперь количество строк совпадает с уникальными значениями _ta_id_ т.е. его можно использовать как первичный ключ для дальнейших очисток

### Исследование  price_range

In [238]:
df.price_range.value_counts()

$$ - $$$    18402
$            6276
$$$$         1423
Name: price_range, dtype: int64

Это ординальный признак, но есть пропуски. Преобразуем в числа.

In [ ]:
dct = {'$$ - $$$': 2, '$': 1, '$$$$': 3}
df.price_range.replace(to_replace=dct)

### Исследование и очистка city

In [239]:
df.city.describe()

count      39980
unique        31
top       London
freq        5757
Name: city, dtype: object

### Исследование и очистка cuisine

Сделаем вспомогательный датафрейм

In [241]:
cuisines = df[["ta_id","cuisine_style"]]

In [242]:
import json
cuisines.cuisine_style.fillna('["Unknown"]', inplace=True)
cuisines.cuisine_style = cuisines.cuisine_style.apply(lambda c: json.loads(c.replace('\'','\"')))

Преобразуем массив кухонь в отдельный ряд на кухню для каждого ресторана

In [243]:
cuisines = cuisines.explode("cuisine_style")
cuisines.describe()

ta_id        cuisine_style
count     104843               104843
unique     39980                  126
top     d7231902  Vegetarian Friendly
freq          21                11183

Число уникальных кухонь - 126, но без учета добавленной Unknown - 125

Самые популярные кухни:

In [244]:
cuisines.cuisine_style.value_counts()


Vegetarian Friendly    11183
European               10056
Unknown                 9279
Mediterranean           6271
Italian                 5963
                       ...  
Salvadoran                 1
Yunnan                     1
Burmese                    1
Latvian                    1
Xinjiang                   1
Name: cuisine_style, Length: 126, dtype: int64

Среднее число кухонь на ресторан:

In [245]:
cuisines.groupby("ta_id").nunique().mean()

ta_id            1.000000
cuisine_style    2.622386
dtype: float64

In [ ]:
Число уникальных 

In [51]:
#df = df.drop([c for c in df.columns if df[c].dtype == 'object'], axis = 1) # Drop non-numbers
#df


Ranking  Rating  Number of Reviews
0       5570.0     3.5              194.0
1       1537.0     4.0               10.0
2        353.0     4.5              688.0
3       3458.0     5.0                3.0
4        621.0     4.0               84.0
...        ...     ...                ...
39995    500.0     4.5               79.0
39996   6341.0     3.5              542.0
39997   1652.0     4.5                4.0
39998    641.0     4.0               70.0
39999   4827.0     3.0              260.0

[40000 rows x 3 columns]

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 40000 entries, 0 to 39999
Data columns (total 3 columns):
 #   Column             Non-Null Count  Dtype  
---  ------             --------------  -----  
 0   Ranking            40000 non-null  float64
 1   Rating             40000 non-null  float64
 2   Number of Reviews  40000 non-null  float64
dtypes: float64(3)
memory usage: 937.6 KB


In [42]:
#df["Number of Reviews"].fillna(df["Number of Reviews"].median(), inplace=True)

# Разбиваем датафрейм на части, необходимые для обучения и тестирования модели

In [44]:
# Х - данные с информацией о ресторанах, у - целевая переменная (рейтинги ресторанов)
#X = df.drop(['Restaurant_id', 'Rating'], axis = 1)
X = df.drop([ 'Rating'], axis = 1)
y = df['Rating']

In [45]:
# Загружаем специальный инструмент для разбивки:
from sklearn.model_selection import train_test_split

In [46]:
# Наборы данных с меткой "train" будут использоваться для обучения модели, "test" - для тестирования.
# Для тестирования мы будем использовать 25% от исходного датасета.
X_train, X_test, y_train, y_test = train_test_split(X, y, test_size=0.25)

# Создаём, обучаем и тестируем модель

In [47]:
# Импортируем необходимые библиотеки:
from sklearn.ensemble import RandomForestRegressor # инструмент для создания и обучения модели
from sklearn import metrics # инструменты для оценки точности модели

In [48]:
# Создаём модель
regr = RandomForestRegressor(n_estimators=100)

# Обучаем модель на тестовом наборе данных
regr.fit(X_train, y_train)

# Используем обученную модель для предсказания рейтинга ресторанов в тестовой выборке.
# Предсказанные значения записываем в переменную y_pred
y_pred = regr.predict(X_test)

In [49]:
# Сравниваем предсказанные значения (y_pred) с реальными (y_test), и смотрим насколько они в среднем отличаются
# Метрика называется Mean Absolute Error (MAE) и показывает среднее отклонение предсказанных значений от фактических.
print('MAE:', metrics.mean_absolute_error(y_test, y_pred))

MAE: 0.43664873607503607
